src/composables/useEmailSubscription.ts

```diff
@@ -1,92 +1,248 @@
+    console.error(e);
+  } finally {
+    isSubscriptionsLoading = false;
   }
+}
+
+async function subscribe(email: string, address: string) {
+  loading.value = true;
+
+  const message = {
+    address,
+    email
+  };
+
+  try {
+    const signature = await signWithAlias(message, SubscribeSchema);
+
+    const response = await fetch(import.meta.env.VITE_ENVELOP_URL, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({
+        params: {
+          email,
+          address,
+          signature
+        },
+        method: 'snapshot.subscribe'
+      })
+    });
+
+    const result = await response.json();
 
-  function setPostSubscribeState(
-    newStatus: Status,
-    level: Level,
-    messageKey?: string
-  ) {
-    status.value = newStatus;
-    postSubscribeLevel.value = level;
-    postSubscribeMessage.value = messageKey
-      ? t(`emailSubscription.postSubscribeMessage.${messageKey}`)
-      : '';
+    if (result.result === 'OK') {
+      setPostSubscribeState(Status.success, Level.success, 'success');
+    } else {
+      setPostSubscribeState(Status.error, Level.warning, 'apiError');
+    }
+  } catch (e) {
+    setPostSubscribeState(Status.error, Level.warning, 'unknownError');
+  } finally {
+    loading.value = false;
   }
+}
+
+async function updateSubscriptions() {
+  if (loading.value) return;
+
+  const address = web3Account.value;
+
+  if (!address) return;
+
+  const params = {
+    address,
+    email: 'dmytro@snapshot.org',
+    subscriptions: apiSubscriptions.value
+  };
+
+  try {
+    const signature = await signWithAlias(params, UpdateSubscriptionsSchema);
+
+    const response = await fetch(import.meta.env.VITE_ENVELOP_URL, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({
+        jsonrpc: '2.0',
+        method: 'snapshot.update',
+        params: {
+          ...params,
+          signature
+        }
+      })
+    });
+
+    const result: JSONRPCSuccess | JSONRPCError = await response.json();
+
+    if ('error' in result) {
+      throw new Error(result.error.message);
+    }
 
-  function reset() {
-    setPostSubscribeState(Status.waiting, Level.info);
+    // TODO: refactor this when BE will be ready for removing email
+    if (shouldRemoveEmail.value && !apiSubscriptions.value.length) {
+      apiSubscriptions.value = [];
+    }
+  } catch (error) {
+    await loadEmailSubscriptions();
+  } finally {
+    loading.value = false;
   }
+}
+
+function setPostSubscribeState(
+  newStatus: Status,
+  level: Level,
+  messageKey?: string
+) {
+  status.value = newStatus;
+  postSubscribeLevel.value = level;
+  postSubscribeMessage.value = messageKey
+    ? t(`emailSubscription.postSubscribeMessage.${messageKey}`)
+    : '';
+}
+
+function reset() {
+  setPostSubscribeState(Status.waiting, Level.info);
+}
+
+export function useEmailSubscription() {
```

Comment:
This doesn't look right. Why are you exporting a composable from the Modal?

Reply:
Oh my bad I see now. But why are you writing all the functions and state outside of the composable function?

Reply:
<img width="686" alt="image" src="https://github.com/snapshot-labs/snapshot/assets/51686767/8d532570-b184-404a-b947-58aa542bff97">

Reply:
There are a couple of reasons why I moved everything out of composable function:
1. I don't need to reinitialize functions and variables each time
2. I use a singleton pattern, so it allows me to share some state among multiple components without extra requests to BE
3. I prefer a flattened code over a nested one. And trying to do it everywhere I can

> > > I get the followoing error after connected with my wallet <img alt="image" width="644" src="https://user-images.githubusercontent.com/51686767/239661669-15992d4c-db41-4bd3-a5f8-b737988c3aa1.png">
> > 
> > 
> > I can suppress the second error(because I have intentionally thrown it [here](https://github.com/snapshot-labs/snapshot/pull/3907/files#diff-7329ba39789105ecc8b4eef5d1b841dc255a78c8b484bed38b8e1c24163a27e9R114)). But the first one comes from BE. Do you propose to change the approach and rewrite BE so it will answer 200 success code, but with an error code?
> 
> My question is more, why is there an error at all after connecting my wallet.

I see. Sam, sorry for the inconvenience. Probably we need to wait until BE for this ticket will be merged. And then I will be able to create a robust prototype/feature.

Reply:
I have looked through `vueuse` library and they also nest all their functions and variables inside a composable. Let me check it again, I will try to rewrite it or will come up with a better explanation of why I need this approach.

Reply:
Yes it makes sense for state that needs to be shared across multiple instances of the composable. I don't think that is the case for all or any of the state in this composable though.